File: backup/ui-styling-improvements-20250913-034719/src/lib/bookingRepository.ts
// Exact replica of Android BookingRepository.kt Firebase structure and logic

import { 
  collection, 
  doc, 
  addDoc, 
  query, 
  where, 
  onSnapshot, 
  orderBy,
  Timestamp,
  getDocs,
  updateDoc
} from 'firebase/firestore';
import { db, auth } from '@/lib/firebase/config';
import { Booking, BookingStatus } from '@/types/booking';

export class BookingRepository {
  private static instance: BookingRepository;
  private bookingsCollection = collection(db, 'bookings');
  private listeners: (() => void)[] = [];

  static getInstance(): BookingRepository {
    if (!BookingRepository.instance) {
      BookingRepository.instance = new BookingRepository();
    }
    return BookingRepository.instance;
  }

  // Create a new booking (matches Android createBooking)
  async createBooking(bookingData: Omit<Booking, 'id' | 'createdAt' | 'updatedAt'>): Promise<string> {
    try {
      const now = Date.now();
      const booking: Omit<Booking, 'id'> = {
        ...bookingData,
        status: BookingStatus.PENDING,
        createdAt: now,
        updatedAt: now
      };

      const docRef = await addDoc(this.bookingsCollection, booking);
      console.log('Booking created with ID:', docRef.id);
      return docRef.id;
    } catch (error) {
      console.error('Error creating booking:', error);
      throw error;
    }
  }

  // Get confirmed bookings for a shop and service (matches Android logic for availability checking)
  async getConfirmedBookingsForShopService(shopId: string, serviceId: string): Promise<Booking[]> {
    try {
      const q = query(
        this.bookingsCollection,
        where('shopId', '==', shopId),
        where('serviceId', '==', serviceId),
        where('status', 'in', [BookingStatus.PENDING, BookingStatus.ACCEPTED]), // PENDING blocks for 24h, ACCEPTED blocks permanently
        orderBy('requestedDate', 'asc')
      );

      const snapshot = await getDocs(q);
      const bookings = snapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      } as Booking));

      console.log(`Found ${bookings.length} confirmed bookings for shop ${shopId}, service ${serviceId}`);
      return bookings;
    } catch (error) {
      console.error('Error fetching confirmed bookings:', error);
      return [];
    }
  }

  // Listen to bookings for current user (matches Android startBookingsListener)
  startUserBookingsListener(callback: (bookings: Booking[]) => void): () => void {
    const currentUser = auth.currentUser;
    if (!currentUser) {
      console.log('No current user, clearing bookings');
      callback([]);
      return () => {};
    }

    console.log('Starting bookings listener for user:', currentUser.uid);

    // Listen for bookings by userId
    const userIdQuery = query(
      this.bookingsCollection,
      where('customerId', '==', currentUser.uid),
      orderBy('createdAt', 'desc')
    );

    const unsubscribeUserId = onSnapshot(userIdQuery, (snapshot) => {
      console.log(`Bookings listener update - ${snapshot.docs.length} bookings by userId`);
      
      const userBookings = snapshot.docs
        .map(doc => {
          try {
            return { id: doc.id, ...doc.data() } as Booking;
          } catch (error) {
            console.error('Error parsing booking:', error);
            return null;
          }
        })
        .filter(Boolean) as Booking[];

      callback(userBookings);
    }, (error) => {
      console.error('Bookings listener error:', error);
    });

    // Also listen for bookings by email if different from userId (Android pattern)
    let unsubscribeEmail = () => {};
    if (currentUser.email) {
      const emailQuery = query(
        this.bookingsCollection,
        where('customerEmail', '==', currentUser.email),
        orderBy('createdAt', 'desc')
      );

      unsubscribeEmail = onSnapshot(emailQuery, (snapshot) => {
        console.log(`Bookings by email listener update - ${snapshot.docs.length} bookings by email`);
        // This would combine with userId bookings in a real implementation
        // For now, we're focusing on the primary userId listener
      }, (error) => {
        console.error('Bookings by email listener error:', error);
      });
    }

    // Return cleanup function
    return () => {
      unsubscribeUserId();
      unsubscribeEmail();
    };
  }

  // Listen to booking requests for shop owner (matches Android pattern)
  startShopBookingsListener(shopId: string, callback: (bookings: Booking[]) => void): () => void {
    console.log('Starting shop bookings listener for shop:', shopId);

    const shopQuery = query(
      this.bookingsCollection,
      where('shopId', '==', shopId),
      orderBy('createdAt', 'desc')
    );

    const unsubscribe = onSnapshot(shopQuery, (snapshot) => {
      console.log(`Shop bookings listener update - ${snapshot.docs.length} bookings for shop`);
      
      const shopBookings = snapshot.docs
        .map(doc => {
          try {
            return { id: doc.id, ...doc.data() } as Booking;
          } catch (error) {
            console.error('Error parsing shop booking:', error);
            return null;
          }
        })
        .filter(Boolean) as Booking[];

      callback(shopBookings);
    }, (error) => {
      console.error('Shop bookings listener error:', error);
    });

    return unsubscribe;
  }

  // Update booking status (for shop owner responses)
  async updateBookingStatus(
    bookingId: string, 
    status: BookingStatus, 
    responseMessage?: string
  ): Promise<void> {
    try {
      const bookingRef = doc(this.bookingsCollection, bookingId);
      const updateData: Partial<Booking> = {
        status,
        updatedAt: Date.now()
      };

      if (responseMessage) {
        updateData.responseMessage = responseMessage;
      }

      await updateDoc(bookingRef, updateData);
      console.log('Booking status updated:', bookingId, status);
    } catch (error) {
      console.error('Error updating booking status:', error);
      throw error;
    }
  }

  // Clean up all listeners
  cleanup(): void {
    this.listeners.forEach(unsubscribe => unsubscribe());
    this.listeners = [];
  }
}

export const bookingRepository = BookingRepository.getInstance();

File: androidAPP/MyHustle/app/src/main/java/com/blueclipse/myhustle/data/repository/BookingRepository.kt
package com.blueclipse.myhustle.data.repository

import android.util.Log
import com.blueclipse.myhustle.data.model.Booking
import com.blueclipse.myhustle.data.model.BookingStatus
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/**
 * Repository for managing booking data in Firebase
 */
class BookingRepository private constructor() {
    
    private val firestore = FirebaseFirestore.getInstance()
    private val bookingsCollection = firestore.collection("bookings")
    
    private val _bookings = MutableStateFlow<List<Booking>>(emptyList())
    val bookings: StateFlow<List<Booking>> = _bookings.asStateFlow()
    
    init {
        // Start listening for current user's bookings
        startBookingsListener()
    }
    
    private fun startBookingsListener() {
        val currentUser = FirebaseAuth.getInstance().currentUser
        if (currentUser != null) {
            Log.d("BookingRepo", "Starting bookings listener for user: ${currentUser.uid}")
            
            // Listen for bookings by userId
            bookingsCollection
                .whereEqualTo("customerId", currentUser.uid)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        Log.e("BookingRepo", "Bookings listener error", error)
                        return@addSnapshotListener
                    }
                    
                    if (snapshot != null) {
                        Log.d("BookingRepo", "Bookings listener update - ${snapshot.documents.size} bookings by userId")
                        val userBookings = snapshot.documents.mapNotNull { document ->
                            try {
                                document.toObject(Booking::class.java)?.copy(id = document.id)
                            } catch (e: Exception) {
                                Log.e("BookingRepo", "Error parsing booking", e)
                                null
                            }
                        }.sortedByDescending { it.createdAt }
                        
                        _bookings.value = userBookings
                    }
                }
            
            // Also listen for bookings by email if different from userId
            if (currentUser.email != null) {
                bookingsCollection
                    .whereEqualTo("customerEmail", currentUser.email)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null) {
                            Log.e("BookingRepo", "Bookings by email listener error", error)
                            return@addSnapshotListener
                        }
                        
                        if (snapshot != null) {
                            Log.d("BookingRepo", "Bookings by email listener update - ${snapshot.documents.size} bookings by email")
                            val emailBookings = snapshot.documents.mapNotNull { document ->
                                try {
                                    document.toObject(Booking::class.java)?.copy(id = document.id)
                                } catch (e: Exception) {
                                    Log.e("BookingRepo", "Error parsing booking by email", e)
                                    null
                                }
                            }.sortedByDescending { it.createdAt }
                            
                            // Combine with existing bookings, avoiding duplicates
                            val currentBookings = _bookings.value
                            val allBookings = (currentBookings + emailBookings)
                                .distinctBy { it.id }
                                .sortedByDescending { it.createdAt }
                            
                            _bookings.value = allBookings
                        }
                    }
            }
        } else {
            Log.d("BookingRepo", "No current user, clearing bookings")
            _bookings.value = emptyList()
        }
    }
    
    /**
     * Manually refresh bookings for current user
     */
    suspend fun refreshUserBookings() {
        val currentUser = FirebaseAuth.getInstance().currentUser
        if (currentUser != null) {
            try {
                // Get bookings by userId
                val userIdBookings = getBookingsForCustomer(currentUser.uid)
                
                // Get bookings by email if available
                val emailBookings = if (currentUser.email != null) {
                    getBookingsForCustomerByEmail(currentUser.email!!)
                } else {
                    emptyList()
                }
                
                // Combine and remove duplicates
                val allBookings = (userIdBookings + emailBookings)
                    .distinctBy { it.id }
                    .sortedByDescending { it.createdAt }
                
                _bookings.value = allBookings
                Log.d("BookingRepo", "Refreshed ${allBookings.size} bookings for current user")
            } catch (e: Exception) {
                Log.e("BookingRepo", "Error refreshing user bookings", e)
            }
        }
    }
    
    /**
     * Get bookings for a customer by email
     */
    suspend fun getBookingsForCustomerByEmail(customerEmail: String): List<Booking> {
        return try {
            val snapshot = bookingsCollection
                .whereEqualTo("customerEmail", customerEmail)
                .get()
                .await()
            
            snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Booking::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt }
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Get bookings for a specific shop owner
     */
    suspend fun getBookingsForShopOwner(shopOwnerId: String): List<Booking> {
        return try {
            val snapshot = bookingsCollection
                .whereEqualTo("shopOwnerId", shopOwnerId)
                .get()
                .await()
            
            snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Booking::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt }
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Get bookings for a specific shop
     */
    suspend fun getBookingsForShop(shopId: String): List<Booking> {
        return try {
            val snapshot = bookingsCollection
                .whereEqualTo("shopId", shopId)
                .get()
                .await()
            
            snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Booking::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt }
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Get bookings for a specific customer
     */
    suspend fun getBookingsForCustomer(customerId: String): List<Booking> {
        return try {
            val snapshot = bookingsCollection
                .whereEqualTo("customerId", customerId)
                .get()
                .await()
            
            snapshot.documents.mapNotNull { document ->
                try {
                    document.toObject(Booking::class.java)?.copy(id = document.id)
                } catch (e: Exception) {
                    null
                }
            }.sortedByDescending { it.createdAt }
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Create a new booking
     */
    suspend fun createBooking(booking: Booking): Result<String> {
        return try {
            val document = bookingsCollection.add(booking).await()
            Result.success(document.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update booking status
     */
    suspend fun updateBookingStatus(
        bookingId: String, 
        status: BookingStatus, 
        responseMessage: String = ""
    ): Result<Unit> {
        return try {
            val updates = mapOf(
                "status" to status,
                "responseMessage" to responseMessage,
                "updatedAt" to System.currentTimeMillis()
            )
            bookingsCollection.document(bookingId).update(updates).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Delete a booking
     */
    suspend fun deleteBooking(bookingId: String): Result<Unit> {
        return try {
            bookingsCollection.document(bookingId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    companion object {
        @Volatile
        private var INSTANCE: BookingRepository? = null
        
        val instance: BookingRepository
            get() = INSTANCE ?: synchronized(this) {
                INSTANCE ?: BookingRepository().also { INSTANCE = it }
            }
    }
}
